### PR TITLE
[llvm-mca][AMDGPU] Retire instructions that have issue carry over correctly

### PR DESCRIPTION
**Pull request by michaelmaitland**
https://github.com/llvm/llvm-project/issues/83775 shows llvm-mca hits sanitizer error in cycleEnd. There was an instruction that takes multiple cycles to issue and is finished executing directly after issue. Prior to this patch, the instruction is retired on the first issue cycle, despite taking multiple cycles to issue.

To fix this, if an instruction takes multiple cycles to issue and is done executing after issue, let updateCarriedOver retire the instruction when it is fully issued.

### CODE REVIEW

File: llvm/lib/MCA/Stages/InOrderIssueStage.cpp [hunk 1]
```diff
@@ -257,8 +257,9 @@ llvm::Error InOrderIssueStage::tryIssue(InstRef &IR) {
   }
 
   // If the instruction has a latency of 0, we need to handle
-  // the execution and retirement now.
-  if (IS.isExecuted()) {
+  // the execution and retirement now. If the instruction is issued in multiple
+  // cycles, we cannot retire until it is finished issuing.
+  if (IS.isExecuted() && !ShouldCarryOver) {
     PRF.onInstructionExecuted(&IS);
```
Comment (michaelmaitland) [inline llvm/lib/MCA/Stages/InOrderIssueStage.cpp:263]: Maybe we need to have `onInstructionExecuted` and `notifyEvent` called in `updateCarriedOver` too?

File: llvm/lib/MCA/Stages/InOrderIssueStage.cpp [hunk 2]
```diff
@@ -329,6 +335,17 @@ void InOrderIssueStage::updateCarriedOver() {
   else
     Bandwidth -= CarryOver;
 
+  // updateIssuedInst did not handle executed if issue had carry over.
+  if (CarriedOver.getInstruction()->isExecuted()) {
+    PRF.onInstructionExecuted(&IS);
+    LSU.onInstructionExecuted(IR);
+    notifyEvent<HWInstructionEvent>(
+        HWInstructionEvent(HWInstructionEvent::Executed, IR));
+    LLVM_DEBUG(dbgs() << "[E] Instruction #" << IR << " is executed\n");
```
Comment (mshockwave) [inline llvm/lib/MCA/Stages/InOrderIssueStage.cpp:344]: Could we use notifyInstructionExecuted here? I think it's doing the same thing

Comment (michaelmaitland) [reply llvm/lib/MCA/Stages/InOrderIssueStage.cpp]: Done!

**Review by arsenm — APPROVED** [review]
Thanks!

**Comment by michaelmaitland** [comment]
It turns out this solution has a problem that I didn't catch when I checked locally. Going to have to rework this patch a bit.

**Comment by michaelmaitland** [comment]
I think the latest fixup solves the problem I was facing. This patch is ready for review.

**Review by arsenm — APPROVED** [review]

**Pull request merged**